Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift

```diff
@@ -259,43 +313,69 @@ public class Mqtt5Client {
                 let continuationCore = ContinuationCore(continuation: continuation)
                 callbackOptions.completion_callback = unsubscribeCompletionCallback
                 callbackOptions.completion_user_data = continuationCore.passRetained()
-                let result = aws_mqtt5_client_unsubscribe(rawValue, unsubscribePacketPointer, &callbackOptions)
-                guard result == AWS_OP_SUCCESS else {
-                    continuationCore.release()
-                    return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                self.rwlock.read {
+                    // validate the client in case close() is called.
+                    guard let rawValue = self.rawValue else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
+                    let result = aws_mqtt5_client_unsubscribe(rawValue, unsubscribePacketPointer, &callbackOptions)
+                    guard result == AWS_OP_SUCCESS else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
                 }
             }
         }
     }
 
+    /// Discard all operations and cleanup the client. It is MANDATORY function to call to release the client.
     public func close() {
-        self.callbackCore.close()
-        aws_mqtt5_client_release(rawValue)
-        rawValue = nil
+        self.rwlock.write {
```

Comment:
Is it safe to call `close()` from a callback or it'll block indefinitely?

Reply:
No worry. The callbacks will be scheduled in an async task, therefore, the lock would not block the thread in that case. 

PublishReceived Example: 
https://github.com/awslabs/aws-crt-swift/blob/64afcfe69b658661cae87a599f0a09d6f20db01b/Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift#L444